File: packages/format/src/av1/codec.ts
import { FormatError } from "../errors.js";
import type { Av1SequenceHeader } from "./sequence-header.js";

export type Av1Codec = `av01.0.${string}.${"08" | "10"}.0.11${0 | 1 | 2 | 3}.01.01.01.0`;

export function av1CodecFromSequence(
  sequence: Readonly<Av1SequenceHeader>
): Av1Codec {
  if (!Number.isInteger(sequence.level) || sequence.level < 0 || sequence.level > 31) {
    throw new FormatError("PROFILE_INVALID", "AV1 level is invalid");
  }
  const level = String(sequence.level).padStart(2, "0");
  const bitDepth = String(sequence.bitDepth).padStart(2, "0") as "08" | "10";
  return `av01.0.${level}${sequence.tier}.${bitDepth}.0.11${sequence.chromaSamplePosition}.01.01.01.0`;
}

export function isAv1Codec(value: unknown): value is Av1Codec {
  return typeof value === "string" &&
    /^av01\.0\.(?:0[0-9]|[12][0-9]|3[01])[MH]\.(?:08|10)\.0\.11[0-3]\.01\.01\.01\.0$/u.test(value);
}

File: flutter/packages/aval_format/lib/src/av1/codec.dart
/// AV1 codec-string derivation and identification.
///
/// Dart port of `packages/format/src/av1/codec.ts`.
library;

import '../errors.dart';
import 'sequence_header.dart';

/// Fully-qualified AV1 codec string,
/// e.g. `"av01.0.00M.08.0.110.01.01.01.0"`.
typedef Av1Codec = String;

Av1Codec av1CodecFromSequence(Av1SequenceHeader sequence) {
  if (sequence.level < 0 || sequence.level > 31) {
    throw FormatError(
      FormatErrorCode.profileInvalid,
      'AV1 level is invalid',
    );
  }
  final level = sequence.level.toString().padLeft(2, '0');
  final bitDepth = sequence.bitDepth.toString().padLeft(2, '0');
  return 'av01.0.$level${sequence.tier}.$bitDepth.0.11${sequence.chromaSamplePosition}.01.01.01.0';
}

final RegExp _av1CodecPattern = RegExp(
  r'^av01\.0\.(?:0[0-9]|[12][0-9]|3[01])[MH]\.(?:08|10)\.0\.11[0-3]\.01\.01\.01\.0$',
);

bool isAv1Codec(Object? value) =>
    value is String && _av1CodecPattern.hasMatch(value);
